tests/model_fields/test_jsonfield.py

```diff
@@ -0,0 +1,606 @@
+from django.db.models import Count, F, Q, Value
+from django.db.models.expressions import RawSQL
+from django.db.models.fields.json import KeyTextTransform, KeyTransform
+from django.db.models.functions import Cast
+from django.db.utils import DatabaseError, IntegrityError
+from django.test import SimpleTestCase, TestCase, skipUnlessDBFeature
+from django.test.utils import CaptureQueriesContext
+
+from .models import JSONModel, NullableJSONModel, OrderedJSONModel
+
+
+@skipIf(connection.vendor == 'oracle', 'Oracle does not support meta ordering.')
+class TestModelMetaOrdering(TestCase):
+    def test_ordering_by_json_field_value(self):
+        OrderedJSONModel.objects.create(value={'b': 2})
+        OrderedJSONModel.objects.create(value={'a': 3})
+        objects = OrderedJSONModel.objects.all()
+        self.assertEqual(objects[0].value, {'a': 3})
+        self.assertEqual(objects[1].value, {'b': 2})
+
+
+class TestDefaultValue(SimpleTestCase):
+    def _set_default(self, value):
+        field = JSONModel._meta.get_field('value')
+        field.default = value
+        return field.check()
+
+    def tearDown(self):
+        self._set_default(dict)
+        return super().tearDown()
+
+    def test_mutable_default_value(self):
+        mutable_values = [
+            ['foo', 'bar', 123], {'foo': 'bar'},
+        ]
+        for value in mutable_values:
+            with self.subTest(value=value):
+                errors = self._set_default(value)
+                self.assertEqual(len(errors), 1)
+                self.assertIn('default should not be a mutable object', errors[0].msg)
+
+    def test_valid_default_value(self):
+        def callable_obj():
+            return {'it': 'works'}
+
+        valid_values = [
+            None, True, 123, 123.45, 'foo bar', ('foo', 'bar', 123), callable_obj,
+        ]
+        for value in valid_values:
+            with self.subTest(value=value):
+                errors = self._set_default(value)
+                self.assertEqual(len(errors), 0)
+
+
+class TestValidation(TestCase):
+    def _set_encoder_decoder(self, encoder, decoder):
+        field = JSONModel._meta.get_field('value')
+        field.encoder, field.decoder = encoder, decoder
+        return field.check()
+
+    @classmethod
+    def setUpTestData(cls):
+        cls.uuid_value = {'uuid': uuid.UUID('{12345678-1234-5678-1234-567812345678}')}
+
+    def tearDown(self):
+        self._set_encoder_decoder(None, None)
+        return super().tearDown()
+
+    def test_validation_error(self):
+        field = models.JSONField()
+        with self.assertRaises(ValidationError) as err:
+            field.clean(self.uuid_value, None)
+        self.assertEqual(err.exception.code, 'invalid')
+        self.assertEqual(err.exception.message % err.exception.params, 'Value must be valid JSON.')
+
+    def test_not_serializable(self):
+        obj = JSONModel(value=self.uuid_value)
+        with transaction.atomic(), self.assertRaises(TypeError) as ex:
```

Comment:
If possible, could you please change this to `assertRaisesMessage` and match the exception message verbatim?

Reply:
https://github.com/python/cpython/pull/4364 introduced some changes to the exception message between Python 3.6 and 3.7 (if I'm not mistaken), that's why I used `assertIn`.

Reply:
A comment there saying that might help in remembering that. :)

Reply:
Okay, done.